src/spi2_pool.rs

```diff
@@ -0,0 +1,211 @@
+use crate::gpio::{AnyOutputPin, InputPin, Output, OutputPin, PinDriver};
+use crate::peripheral::{Peripheral, PeripheralRef};
+use crate::spi2::{config, SpiBusMasterDriver, SpiMasterDriver};
+
+use std::borrow::Borrow;
+use std::collections::HashMap;
+use std::marker::PhantomData;
+
+use core::ptr;
+
+use esp_idf_sys::*;
+
+// TODO -> custom error that wraps around esperror
+// cs default high low setting -
+// needed to makte SpiBusMasterDriver fields pub because different module
+// needed to make SpiBusMasterDriver flush fn pub
+
+// performance measurment on esp32c3 -> 
+// debug build: 50 uS deley between cs low -> first clk out
+// release build : 15 uS deley between cs low -> first cll out
+
+// debug build: 4uS delay between last clk out -> cs pin high
+// release build: 3uS delay between last clk out -> cs pin high
+
+#[cfg(not(esp32c3))]
+const ESP_MAX_SPI_DEVICES: usize = 3;
+#[cfg(esp32c3)]
+const ESP_MAX_SPI_DEVICES: usize = 6;
+
+pub struct SpiConfigPool<'d, T> {
+    shared_handles: HashMap<u32, spi_device_handle_t>,
+    master: T,
+    _p: PhantomData<&'d ()>,
+}
+
+impl<'d, T> SpiConfigPool<'d, T> {
+    pub fn new(master: T, device_configs: HashMap<u32, config::Config>) -> Result<Self, EspError>
+    where
+        T: Borrow<SpiMasterDriver<'d>> + 'd,
```

Comment:
This should be `BorrowMut` as you want an exclusive ref to the driver for pooling.
```suggestion
        T: BorrowMut<SpiMasterDriver<'d>> + 'd,
```

Reply:
yes and no ? :smile: technically you could have a subset of devices that uses pooling and some that are plain EspSpiDevices. i don't require it exclusively. like 2 normal ones and 6 devices that share one handle. with this you could use some devices with hardware cs and the rest with software cs no ?

Reply:
Sure but how does the `SpiConfigPool` know how many devices are available for pooling then? Especially since this number could change at runtime.

Reply:
> Sure but how does the `SpiConfigPool` know how many devices are available for pooling then? Especially since this number could change at runtime.

how many configs are available uses the same restriction as creating normal EspSpiDevices. you crate an SpiConfigPool with an set of configs. one should be able to change an config but not alter the numbers of configs. the user must ensure that in total all numbers of EspSpiDevices in parallel with the number of configs given to the SpiConfigPool creation does not exide 3 / 6 devices. you than get en EspError if you use more than that. but yeah would be nice to have that sorted out at compile time.
i saw some crazy shit with constant generics combined with FnOnce impl where one can construct functions that are only callable N times and that is checked at compile time !!!